File: rxui-sample-java/src/main/java/com/artemzin/rxui/sample/java/MainView.java
package com.artemzin.rxui.sample.java;


import io.reactivex.Observable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;

// Observable <-> Observable.
interface MainView {
    // Produces.
    Observable<String> login();
    Observable<String> password();
    Observable<Object> signInClicks();

    // Consumes.
    Function<Observable<Object>, Disposable> signInEnable();
    Function<Observable<Object>, Disposable> signInDisable();
    Function<Observable<AuthService.Success>, Disposable> signInSuccess();
    Function<Observable<AuthService.Failure>, Disposable> signInFailure();
}

File: rxui-sample-kotlin/src/main/kotlin/com/artemzin/rxui/sample/kotlin/MainView.kt
package com.artemzin.rxui.sample.kotlin

import com.artemzin.rxui.sample.kotlin.AuthService.Response.Failure
import com.artemzin.rxui.sample.kotlin.AuthService.Response.Success
import io.reactivex.Observable
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function

// Observable <-> Observable.
interface MainView {
    // Produces.
    val login: Observable<String>
    val password: Observable<String>
    val signInClicks: Observable<Unit>

    // Consumes.
    val signInEnable: Function<Observable<Unit>, Disposable>
    val signInDisable: Function<Observable<Unit>, Disposable>
    val signInSuccess: Function<Observable<Success>, Disposable>
    val signInFailure: Function<Observable<Failure>, Disposable>
}
